bitcoin/src/merkle_tree/mod.rs

```diff
@@ -39,11 +87,7 @@ pub use self::block::{MerkleBlock, MerkleBlockError, PartialMerkleTree};
 /// - `None` if `hashes` is empty. The merkle root of an empty tree of hashes is undefined.
 /// - `Some(hash)` if `hashes` contains one element. A single hash is by definition the merkle root.
 /// - `Some(merkle_root)` if length of `hashes` is greater than one.
-pub fn calculate_root_inline<T>(hashes: &mut [T]) -> Option<T>
-where
-    T: Hash + Encodable,
-    <T as Hash>::Engine: Write,
-{
+pub fn calculate_root_inline<T: MerkleNode>(hashes: &mut [T]) -> Option<T> {
```

Comment:
This moves the problem I described (that I thought needed `unsafe` to solve) onto the user.

Said another way, what is the point of providing this function now, presumably the caller has a vector of `Txid`s and needs to convert them to `MerkleNode` implementors?

Reply:
We could take the `pub` away if it worries you. It is only used within the library as a helper function to implement `calculate_root`, and it exists to theoretically allow people to implement `calculate_root` without any allocations.

I doubt it has ever been actually used by anybody else in the history of this library.

Reply:
BTW the algorithm implemented here is super memory-wasteful and bizarrely combines recursion with an explicit Vec, giving us the worst of both worlds (and spread across 3 functions for some reason). I'll do a followup PR that fixes this since I'm looking at it.

Reply:
Nah, I'll just do it here. If I'm breaking the API I should go all the way. Added two commits.

Reply:
Yeah, if we could make it private I'd like that. If someone wants it public later we can easily change it back, right.